### PR TITLE
Preserve Context in LoginScanner socket calls, fixes #4723

### LINKED ISSUE #4723: Rex::Socket.create calls should always pass in Context
Without the Context element, these sockets can act in strange ways:
Examples of broken code:

* https://github.com/rapid7/metasploit-framework/blob/master/lib/metasploit/framework/tcp/client.rb#L84

* https://github.com/rapid7/metasploit-framework/blob/master/lib/metasploit/framework/login_scanner/snmp.rb#L41

**Update**: Switchboard is unaffected, but there is various other logic scattered around that depends on the Context being set.

### PR DESCRIPTION
**Pull request by hdm**
During the LoginScanner rework, a bunch of code was pulled from mixins and moved into the LoginScanner namespace, but in doing so, did not pass the Context hash into the Rex::Socket.create calls. Without the Context hash, any code using the socket event handlers are unable to determine what module created the socket. This is a fairly obscure feature, but has been consistently implemented since the first version of Metasploit Framework 3.x.

To demonstrate the issue, a modified version of the socket_logger plugin has been included in this PR. 

Using the new code:

```
msf> use auxiliary/scanner/ftp/ftp_login
msf auxiliary(ftp_login) > set RHOSTS ftp.openbsd.org
msf auxiliary(ftp_login) > set USERNAME ftp
msf auxiliary(ftp_login) > set PASSWORD hello@metasploit.com

msf auxiliary(ftp_login) > load socket_logger
[*] Successfully loaded plugin: socket_logger

msf auxiliary(ftp_login) > run
[*] 129.128.5.191:21 - Starting FTP login sweep
[+] 129.128.5.191:21 - LOGIN SUCCESSFUL: ftp:hello@metasploit.com
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed

msf auxiliary(ftp_login) > ls -la /tmp/socket_*
[*] exec: ls -la /tmp/socket_*
-rw-r--r-- 1 hdm hdm 3149 Feb  9 12:30 /tmp/socket_0.log

msf auxiliary(ftp_login) > cat /tmp/socket_0.log
[*] exec: cat /tmp/socket_0.log

Socket created at 2015-02-09 12:30:42 -0600 (1423506642)
Info: TCP 0.0.0.0:0 > 129.128.5.191:21 via auxiliary/scanner/ftp/ftp_login

READ    47  MjIwIG9wZW5ic2Quc3J2LnVhbGJlcnRhLmNhIEZUUCBzZXJ2ZXIgcmVhZHkuDQo=
WRITE   10  VVNFUiBmdHANCg==
READ    58  MzMxIEd1ZXN0IGxvZ2luIG9rLCBzZW5kIHlvdXIgZW1haWwgYWRkcmVzcyBhcyBwYXNzd29yZC4NCg==
WRITE   27  UEFTUyBoZWxsb0BtZXRhc3Bsb2l0LmNvbQ0K
READ    65  MjMwLSAgIFdlbGNvbWUgdG8gZnRwLm9wZW5ic2Qub3JnIGF0IHRoZSBVbml2ZXJzaXR5IG9mIEFsYmVydGEgDQo=
READ    1448    MjMwLSAgIGluIEVkbW9udG9uLCBBbGJlcnRhLCBDYW5hZGEuDQoyMzAtICAgRm9yIG90aGVyIG1pcnJvciBzaXRlcyB2aXNpdCBodHRwOi8vd3d3Lm9wZW5ic2Qub3JnL2Z0cC5odG1sDQoyMzAtIA0KMjMwLSAgICAgICAgIF9fX19fICAgICAgICAgICAgICAgICBfX19fICAgX19fX18gX19fX18NCjIzMC0gICAgICAgIC8gX19fIFwgICAgICAgICAgICAgICB8ICBfIFwgLyBfX19ffCAgX18gXA0KMjMwLSAgICAgICAvIC8gIC8gL19fXyAgX19fICBfX19fIHwgfF8pIHwgKF9fXyB8IHwgIHwgfA0KMjMwLSAgICAgIC8gLyAgLyAvIF9fIFwvIF8gXC8gX18gXHwgIF8gPCBcX19fIFx8IHwgIHwgfA0KMjMwLSAgICAgLyAvX18vIC8gL18vIC8gIF9fLyAvIC8gL3wgfF8pIHxfX19fKSB8IHxfX3wgfA0KMjMwLSAgICAgXF9fX19fLyAuX19fL1xfX18vXy8gL18vIHxfX19fL3xfX19fXy98X19fX18vDQoyMzAtICAgICAgICAgIC9fLw0KMjMwLSAgICAgICAgICAgICAgICAgfCAgICAuICAgICAgICAgICAgVGhlIHByb2FjdGl2ZWx5IHNlY3VyZSBVbml4LWxpa2UNCjIzMC0gICAgICAgICAgICAgLiAgIHxMICAvfCAgIC4gICAgICAgIE9wZXJhdGluZyBTeXN0ZW0uDQoyMzAtICAgICAgICAgXyAuIHxcIF98IFwtLSsuXy98IC4gICAgICBQbGVhc2UgdmlzaXQgdGhlIE9wZW5CU0Qgd2ViIHNpdGUNCjIzMC0gICAgICAgIC8gfHxcfCBZIEogICkgICAvIHwvfCAuLyAgICAgIGF0IGh0dHA6Ly93d3cub3BlbmJzZC5vcmcvDQoyMzAtICAgICAgIEogIHwpJyggfCAgICAgICAgYCBGYC4nLw0KMjMwLSAgICAgLTx8ICBGICAgICAgICAgX18gICAgIC4tPCAgICAgQWxsIHRyYW5zZmVycyBhcmUgbG9nZ2VkLCBpZiB5b3UgZG9uJ3QNCjIzMC0gICAgICAgfCAvICAgICAgIC4tJy4gYC4gIC8tLiBMX19fIGxpa2UgdGhpcyBwb2xpY3ksIGRpc2Nvbm5lY3Qgbm93IQ0KMjMwLSAgICAgICBKIFwgICAgICA8ICAgIFwgIHwgfCBPXHwuLScNCjIzMC0gICAgIF9KIFwgIC4tICAgIFwvIE8gfCB8IFwgIHxGDQoyMzAtICAgICctRiAgLTxfLiAgICAgXCAgIC4tJyAgYC0nIExfXyBPcGVuQlNEIGlzIGF2YWlsYWJsZSBmb3Igb3JkZXIhDQoyMzAtICAgX19KICBfICAgXy4gICAgID4tJyAgKS5fLiAgIHwtJyBZb3UgY2FuIG9yZGVyIE9wZW5CU0QgQ0QncyBmcm9tIA0KMjMwLSAgIGAtfC4nICAgL18uICAgICAgICAgICBcX3wgICBGICAgaHR0cDovL3d3dy5vcGVuYnNkLm9yZy9vcmRlcnMuaHRtbC4NCjIzMC0gICAgIC8uLSAgIC4gICAgICAgICAgICAgICAgXy48ICAgIENEIHNhbGVzIGFyZSB2ZXJ5IGltcG9ydGFudCB0byBzdXBwb3J0DQoyMzAtICAgIC8nICAgIC8uJyAgICAgICAgICAgICAuJyAgYFwgICB0aGUgY29udGludWVkIGRldmVsb3BtZW50IG9mIHRoZSBwcm9qZWN0Lg0KMjMwLSAgICAgL0wgIC8nICAgfC8gICAgICBfLi0nLVwNCjIzMC0gICAgLydKICAgICAgIF8=
READ    549 X18uLS0tJ1x8DQoyMzAtICAgICAgfFwgIC4tLScgViAgfCBgLiBgDQoyMzAtICAgICAgfC9gLiBgLS4gICAgIGAuXykNCjIzMC0gICAgICAgICAvIC4tLlwNCjIzMC0gICBWSyAgICBcICggIGBcDQoyMzAtICAgICAgICAgIGAuXA0KMjMwLSAgDQoyMzAtICAgKkRPIE5PVCogbWlycm9yIG9wZW5ic2QgZnJvbSB0aGlzIHNpdGUhIHVzZSBvbmUgb2YgdGhlDQoyMzAtICAic2Vjb25kIGxldmVsIG1pcnJvcnMiIGxpc3RlZCBhdCBodHRwOi8vd3d3Lm9wZW5ic2Qub3JnL2Z0cC5odG1sDQoyMzAtICBpbnN0ZWFkIG9mIHRoaXMgc2l0ZS4gIElmIHlvdSBtaXJyb3IgZnJvbSB0aGlzIHNpdGUgeW91IHdpbGwgbG9zZSANCjIzMC0gIGFjY2VzcyB0byBpdC4NCjIzMC0gDQoyMzAtICBFLW1haWwgY29tbWVudHMsIHF1ZXN0aW9ucywgdHJvdWJsZSByZXBvcnRzLCBhbmQgY29tcGxhaW50cw0KMjMwLSAgdG8gYmVja0BvcGVuYnNkLm9yZy4gIFBsZWFzZSBkcml2ZSBzYWZlbHkuDQoyMzAtIA0KMjMwIEd1ZXN0IGxvZ2luIG9rLCBhY2Nlc3MgcmVzdHJpY3Rpb25zIGFwcGx5Lg0K
```

Switching back to master, the socket log isn't even being created:

```
$ rm -f /tmp/socket_[0-9]
$ git checkout master
$ ./msfconsole
msf> use auxiliary/scanner/ftp/ftp_login
msf auxiliary(ftp_login) > set RHOSTS ftp.openbsd.org
msf auxiliary(ftp_login) > set USERNAME ftp
msf auxiliary(ftp_login) > set PASSWORD hello@metasploit.com

msf auxiliary(ftp_login) > load socket_logger
[*] Successfully loaded plugin: socket_logger

msf auxiliary(ftp_login) > run
[*] 129.128.5.191:21 - Starting FTP login sweep
[+] 129.128.5.191:21 - LOGIN SUCCESSFUL: ftp:hello@metasploit.com
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed

msf auxiliary(ftp_login) > ls -la /tmp/socket_*
[*] exec: ls -la /tmp/socket_*
ls: cannot access /tmp/socket_*: No such file or directory
```

In other words, without passing Context into each Rex::Socket call, the socket event handler in Framework doesn't see and has no ability to hook socket operations. This is a regression introduced with the new LoginScanner code and was not a problem prior to this change.

### CODE REVIEW

**Comment by hmoore-r7** [comment]
The ```ips_filter``` plugin is also affected by this regression.

**Comment by todb-r7** [comment]
The socket_log demonstration is super helpful, thx. This is much clearer now.
